### PR TITLE
add external margin

### PR DESCRIPTION
**Pull request by rachitgarg91**
This adds a new envvar (NVTE_EXT_MARGIN_SM) that adds an additional carveout to all the gemms for DP.

### CODE REVIEW

File: transformer_engine/pytorch/csrc/ts_fp8_op.cpp [hunk 1]
```diff
@@ -316,6 +318,11 @@ at::Tensor te_gemm_ts(at::Tensor A,
   bool accumulate_arg = static_cast<bool>(accumulate);
   bool use_split_accumulator_arg = static_cast<bool>(use_split_accumulator);
 
+  // Set an external SM Margin to all the GEMMs. This comes in handy when DP is overlapped with GEMMs
+  cudaDeviceProp prop;
+  cudaGetDeviceProperties(&prop, 0);
+  int num_math_sms = prop.multiProcessorCount - transformer_engine::getenv<int>("NVTE_EXT_MARGIN_SM", 0);
```
Comment (timmoon10) [inline transformer_engine/pytorch/csrc/ts_fp8_op.cpp:324]: This doesn't seem to be used. Do we need to pass it to `te_gemm`?

Comment (rachitgarg91) [reply transformer_engine/pytorch/csrc/ts_fp8_op.cpp]: Yes, I missed that in this branch, fixed

**Review by ksivaman — CHANGES_REQUESTED** [review]
@rachitgarg91 Could you rebase your branch to a clean main. Currently it includes some stray commits.

**Review by ksivaman — APPROVED** [review]
LGTM

**Pull request merged**